lib/spack/spack/test/concretize.py

```diff
@@ -3047,6 +3048,49 @@ def test_spec_filters(specs, include, exclude, expected):
     assert f.selected_specs() == expected
 
 
+@pytest.mark.only_clingo("Feature not implemented in the original concretizer")
+def test_branch_based_versions_pin_to_commits(
+    mock_git_version_info, database, mock_packages, monkeypatch, do_not_check_runtimes_on_reuse
+):
+    repo_path, filename, commits = mock_git_version_info
+    monkeypatch.setattr(
+        spack.package_base.PackageBase, "git", pathlib.Path(repo_path).as_uri(), raising=False
+    )
+
+    spec = Spec("git-test-commit@main").concretized()
+    # assure it is not a StandardVersion post solve
+    assert isinstance(spec.versions.concrete, GitVersion)
+    # last main commit was 3'rd in the list (see mock_git_version_info)
+    assert spec.format("{version}") == f"git.{commits[2]}=main"
+
+
+@pytest.mark.only_clingo("Feature not implemented in the original concretizer")
+def test_versions_with_custom_git_branch_based_versions_pin_to_commits(
+    mock_git_version_info, database, mock_packages, monkeypatch, do_not_check_runtimes_on_reuse
+):
+    import spack.pkg.builtin.mock.version_test_pkg as vtp
+
+    repo_path, filename, commits = mock_git_version_info
+
+    version = Version("develop")
+    new_version_attrs = vtp.VersionTestPkg.versions[version]
+
+    new_version_attrs["git"] = pathlib.Path(repo_path).as_uri()
+    new_version_attrs["branch"] = "main"
+
+    patch_versions = {version: new_version_attrs}
+
+    monkeypatch.setattr(vtp.VersionTestPkg, "versions", patch_versions)
```

Comment:
This is not working. It seems to be using the original git ref posted in the `version-test-pkg` develop branch. Welcome suggestions on how to patch this to the local git repo constructed for unit-testing.